service/gcs/bridge/bridge.go

```diff
@@ -14,385 +16,563 @@ import (
-			logrus.Info("received from HCS: ComputeSystemResizeConsoleV1")
-			response, err = b.resizeConsole(message)
+			logrus.Infof("bridge: read message '%s'\n", message)
+			requestChan <- &Request{header, message}
+		}
+	}()
+	// process each bridge request async and create the response writer
+	go func() {
+		for req := range requestChan {
+			go func(r *Request) {
+				wr := &requestResponseWriter{
+					header: &prot.MessageHeader{
+						Type: prot.GetResponseIdentifier(r.Header.Type),
+						ID:   r.Header.ID,
+					},
+					respChan: b.responseChan,
+				}
+				b.Handler.ServeMsg(wr, r)
+				if !wr.respWritten {
+					logrus.Errorf("bridge: request: ID: %d, Type: %d failed to write a response.\n", r.Header.ID, r.Header.Type)
+				}
+			}(req)
+		}
+	}()
+	// process each bridge response sync this is valid for request/response and publish workflows
+	go func() {
+		for resp := range b.responseChan {
+			responseBytes, err := json.Marshal(resp.response)
 			if err != nil {
-				logrus.Error(err)
+				responseErrChan <- errors.Wrapf(err, "bridge: failed to marshal JSON for response \"%v\"", resp.response)
+				continue
 			}
-		case prot.ComputeSystemModifySettingsV1:
-			logrus.Info("received from HCS: ComputeSystemModifySettingsV1")
-			response, err = b.modifySettings(message)
-			if err != nil {
-				logrus.Error(err)
+			resp.header.Size = uint32(len(responseBytes) + prot.MessageHeaderSize)
+			if err := binary.Write(b.commandConn, binary.LittleEndian, resp.header); err != nil {
+				responseErrChan <- errors.Wrap(err, "bridge: failed writing message header")
+				continue
 			}
-		default:
-			// TODO: Should a response be returned in this case?
-			logrus.Error(errors.Errorf("received invalid header type code from HCS: 0x%x", header.Type))
-		}
 
-		// Set the error fields on the response if an error was encountered.
-		if err != nil {
-			switch response := response.(type) {
-			case *prot.MessageResponseBase:
-				b.setErrorForResponseBase(response, err)
-			case *prot.ContainerCreateResponse:
-				b.setErrorForResponseBase(response.MessageResponseBase, err)
-			case *prot.ContainerExecuteProcessResponse:
-				b.setErrorForResponseBase(response.MessageResponseBase, err)
-			case *prot.ContainerWaitForProcessResponse:
-				b.setErrorForResponseBase(response.MessageResponseBase, err)
-			case *prot.ContainerGetPropertiesResponse:
-				b.setErrorForResponseBase(response.MessageResponseBase, err)
-			default:
-				// TODO: Should this error be handled better?
-				return errors.Errorf("invalid response type: %T", response)
+			if _, err := b.commandConn.Write(responseBytes); err != nil {
+				responseErrChan <- errors.Wrap(err, "bridge: failed writing message payload")
+				continue
 			}
+			logrus.Infof("bridge: response sent: '%s' to HCS\n", responseBytes)
 		}
+	}()
+	// If we get any errors. We return from Listen and shutdown the bridge connection.
+	// TODO: We should likely just log a generic response and keep listening?
```

Comment:
I think this falls under the fail-fast discussion we had before. There are probably some specific connection errors it makes sense to retry, but for most errors (the connection is closed, the HCS sends an invalid message, we send an invalid message), it seems like something is probably pretty wrong/unrecoverable, and we should just exit rather than trying to recover. That isn't to say we shouldn't not retry/try to recover on certain errors, but I feel like the default behavior should just be to exit. Does that make sense?